python/ray/external_storage.py

```diff
@@ -76,37 +165,39 @@ class FileSystemStorage(ExternalStorage):
 
     def __init__(self, directory_path):
         self.directory_path = directory_path
-        self.prefix = "ray_spilled_object_"
+        self.prefix = DEFAULT_OBJECT_PREFIX
         os.makedirs(self.directory_path, exist_ok=True)
         if not os.path.exists(self.directory_path):
             raise ValueError("The given directory path to store objects, "
                              f"{self.directory_path}, could not be created.")
 
-    def spill_objects(self, object_refs):
-        keys = []
-        ray_object_pairs = self._get_objects_from_store(object_refs)
-        for ref, (buf, metadata) in zip(object_refs, ray_object_pairs):
-            filename = self.prefix + ref.hex()
-            with open(os.path.join(self.directory_path, filename), "wb") as f:
-                metadata_len = len(metadata)
-                buf_len = len(buf)
-                f.write(metadata_len.to_bytes(8, byteorder="little"))
-                f.write(buf_len.to_bytes(8, byteorder="little"))
-                f.write(metadata)
-                f.write(memoryview(buf))
-            keys.append(filename.encode())
-        return keys
-
-    def restore_spilled_objects(self, keys):
-        for k in keys:
-            filename = k.decode()
-            ref = ray.ObjectRef(bytes.fromhex(filename[len(self.prefix):]))
-            with open(os.path.join(self.directory_path, filename), "rb") as f:
+    def spill_objects(self, object_refs) -> List[str]:
+        if len(object_refs) == 0:
+            return []
+        # Always use the first object ref as a key when fusioning objects.
+        first_ref = object_refs[0]
+        filename = f"{self.prefix}-{first_ref.hex()}-multi-{len(object_refs)}"
+        url = f"{os.path.join(self.directory_path, filename)}"
+        with open(url, "wb") as f:
+            return self._write_multiple_objects(f, object_refs, url)
+
+    def restore_spilled_objects(self, object_refs: List[ObjectRef],
+                                url_with_offset_list: List[str]):
+        for i in range(len(object_refs)):
+            object_ref = object_refs[i]
+            url_with_offset = url_with_offset_list[i].decode()
+            # Retrieve the information needed.
+            parsed_result = parse_url_with_offset(url_with_offset)
+            base_url = parsed_result.base_url
+            offset = parsed_result.offset
```

Comment:
We seem to be ignoring the size here. How about checking the size matches the recorded metadata and raising an error if they don't match?

Reply:
Sounds good. Will do!